### ISSUE TITLE
new multiple sources doesn't work with helm-secrets

### ISSUE THREAD

**Issue by brandonguigo** [argoproj/argo-cd#11866]
Checklist:

* [X] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [ ] I've included steps to reproduce the bug.
* [ ] I've pasted the output of `argocd version`.

**Describe the bug**

When using as values files in helm application with secrets, using the "secrets+age-import:///age-secret-key/keys.txt" file uri, the ref is not replaced with the file path. 

**To Reproduce**

```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: namespace-config
spec:
  project: project
  destination:
    name: in-cluster
    namespace: staging
  source:
    - repoURL: repo2
      targetRevision: main
      ref: config
    - repoURL: repo # Can point to either a Helm chart repo or a git repo.
      targetRevision: HEAD  # For Helm, this refers to the chart version.
      path: path # This has no meaning for Helm charts pulled directly from a Helm repo instead of git.
      helm:
        valueFiles:
          - values.yaml
          - $config/namespace-config/default/values.namespace.yaml
          - secrets+age-import:///age-secret-key/keys.txt?$config/namespace-config/default/secrets.namespace.yaml
          - $config/namespace-config/staging/values.namespace.yaml
          - secrets+age-import:///age-secret-key/keys.txt?$config/namespace-config/staging/secrets.namespace.yaml
      ignoreMissingValueFiles: false
  syncPolicy:
    automated: # automated sync by default retries failed attempts 5 times with following delays between attempts ( 5s, 10s, 20s, 40s, 80s ); retry controlled using `retry` field.
      prune: true # Specifies if resources should be pruned during auto-syncing ( false by default ).
      selfHeal: true # Specifies if partial app sync should be executed when resources are changed only in target Kubernetes cluster and no git change detected ( false by default ).
      allowEmpty: false # Allows deleting all application resources during automatic syncing ( false by default ).
    syncOptions:     # Sync options which modifies sync behavior
      - Validate=false # disables resource validation (equivalent to 'kubectl apply --validate=false') ( true by default ).
      - CreateNamespace=true # Namespace Auto-Creation ensures that namespace specified as the application destination exists in the destination cluster.
      - PrunePropagationPolicy=foreground # Supported policies are background, foreground and orphan.
      - PruneLast=true # Allow the ability for resource pruning to happen as a final, implicit wave of a sync operation
    # The retry feature is available since v1.7
    retry:
      limit: 5 # number of failed sync attempt retries; unlimited number of attempts if less than 0
      backoff:
        duration: 5s # the amount to back off. Default unit is seconds, but could also be a duration (e.g. "2m", "1h")
        factor: 2 # a factor to multiply the base duration after each failed retry
        maxDuration: 3m # the maximum amount of time allowed for the backoff strategy
  # RevisionHistoryLimit limits the number of items kept in the application's revision history, which is used for
  # informational purposes as well as for rollbacks to previous versions. This should only be changed in exceptional
  # circumstances. Setting to zero will store no history. This will reduce storage used. Increasing will increase the
  # space used to store the history, so we do not recommend increasing it.
  revisionHistoryLimit: 1
```

**Expected behavior**

ArgoCD should replace $config by the directory path

**Screenshots**

**Version**

```shell
Paste the output from `argocd version` here.
2.6.0-rc1 (i have no ingress on the argocd deployment so argocd version don't work)
```

**Logs**

```
Paste any relevant application logs here.
```

**Comment by morey-tech** [comment]
To clarify the example to reproduce, you have `spec.source` (singular) when it should be `spec.sources` (plural). Is that the issue or only in the example?

**Comment by brandonguigo** [comment]
Oh sorry, it's only in the examples, I have sources inside my real yaml file

**Comment by brandonguigo** [comment]
@morey-tech Hey, so just to clarify, it seems that the fix will not be included in the 2.6 release ? 
I'm stuck with submodules everytime i want to use secrets for now ?

**Comment by morey-tech** [comment]
> @morey-tech Hey, so just to clarify, it seems that the fix will not be included in the 2.6 release ? I'm stuck with submodules everytime i want to use secrets for now ?

Based on https://github.com/argoproj/argo-cd/pull/11966#issuecomment-1386387547, unfortunately, it looks that way. As [a beta feature](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md#beta), there may be edge cases that aren't supported yet, and Helm Secrets happen to be a victim of that. Unless someone has time to figure out the solution and properly test it before the 2.6 release on 2023-02-06, it'll have to wait for [2.7 around 2023-04-10](https://github.com/orgs/argoproj/projects/25) or there is "some hope for including it in a patch release"

**Comment by aoz-turk** [comment]
Hi,

We want to use helm-secrets with sops backend + multi sources app. From the following documentation, we understand that the limitation lives on Argo CD side. 

https://github.com/jkroepke/helm-secrets/wiki/ArgoCD-Integration#multi-source-application-support-beta

Should we wait for 2.7 release?

**Comment by jkroepke** [comment]
Ping @todaywasawesome, greetings from KubeCon.

**Comment by jkroepke** [comment]
It seems like that this issue can be resolved by #12508

@ishitasequeira Did you see any chance that the PR will be merged into 2.7?

**Comment by ishitasequeira** [comment]
@jkroepke Thanks for reaching out.. I have reviewed the PR!! Looking for another eyes from the approvers to get it merged.

**Comment by jkroepke** [comment]
@egoist777 #12508 is not merged, nothing changed in 2.7.0 in terms of this.

**Comment by egoist777** [comment]
@jkroepke I see...
Can you please suggest any workaround for this ? Thanks in advance.

**Comment by tol777900** [comment]
> @egoist777 #12508 is not merged, nothing changed in 2.7.0 in terms of this.

Any updates when this will get merged ?

**Comment by harrywm** [comment]
> Any updates when this will get merged ?

Seems it's just waiting for an extra reviewer, similar to pre-2.7. Any help testing the feature of course helps get things through quicker!

**Comment by madispuk** [comment]
Any updates?

**Comment by MohammedNoureldin** [comment]
Soon 2.9.0 will be out. Is there any plans for this? Being able to use encrypted values.yaml is curcial IMHO.

I am a bit confused, because helm-secrets people mention here that there is a limitation, and this is shown here in this issue. However, there is an example shown which supposes that it should work. I am a but confused.

I am talking about this:

https://github.com/jkroepke/helm-secrets/wiki/ArgoCD-Integration

```
[Multi-Source Application Support [BETA]](https://github.com/jkroepke/helm-secrets/wiki/ArgoCD-Integration#multi-source-application-support-beta)
ArgoCD has limited supported for helm-secrets and Multi-Source application. sops backend is not tested yet.

References:

https://github.com/argoproj/argo-cd/issues/11866
https://github.com/argoproj/argo-cd/pull/11966
On ArgoCD 2.6.x, helm-secrets isn't supported in Multi-Source application, because the source reference, e.g.: $ref needs to be at the beginn of a string. This is in conflict with helm-secrets, since the string needs to beginn with secrets://. On top, ArgoCD do not resolve references in URLs.

HELM_SECRETS_VALUES_ALLOW_ABSOLUTE_PATH must be set to true, since ArgoCD pass value files with absolute file path.

Ensure that the env HELM_SECRETS_WRAPPER_ENABLED=true (default false) and HELM_SECRETS_VALUES_ALLOW_ABSOLUTE_PATH=true is set on the argocd-repo-server. Please ensure you are following the lastest installation instructions (updated on 2023-03-03).
```

Could any one tell me what is the current status?

**Comment by jkroepke** [comment]
@MohammedNoureldin 

PR #12508  is likely something that we are looking for, but its getting staled

**Comment by crenshaw-dev** [comment]
@MohammedNoureldin I'm not sure how that workaround is supposed to work or why it might fail... Maybe they have a custom Argo CD image which supports those env vars?

Re: #12508, it's not stale, I just haven't had time to give it a final review and get it merged yet.

**Comment by MohammedNoureldin** [comment]
Thank you, @crenshaw-dev and @jkroepke.

@crenshaw-dev I know you are busy, I am just wondering, as the others do in that PR, if we have a chance in getting this PR in 2.9.

**Comment by l-maia** [comment]
I think that env is for helm-secrets that would by default reject an absolute path, not argo.

Meanwhile 1 week into testing argocd I already hit this limitation too, only because the substitution assumes that the string must start with $ref.

The simplest method for me would be to allow ref substitution (in helm plugin) using both a file  path or an URI style reference, without touching the schema to allow for arbitrary plugins (secrets://).

**Comment by jkroepke** [comment]
As plugin maintainer, i'm open to any reachable solution.

**Comment by jkroepke** [comment]
Whats is the issue with allow to substitute the $ref after the protocol handler `secrets//$ref/`? That would be the best user expericence.

**Comment by harrywm** [comment]
> @jkroepke I see... Can you please suggest any workaround for this ? Thanks in advance.

@egoist777 We're currently using a Helm SubChart method as a workaround, and we actually prefer it to the proposed Multi-App solution. Though, of course, won't work for all cases.

Here's a gist that covers our setup.
https://gist.github.com/harrywm/6a80b6dc355f20595b002734559d4a15

**Comment by l-maia** [comment]
> Yeah I wrote up a small proposal on another issue about how we could do ref substitution on secret values. At the time it seemed gnarlier than the solution of sharing arbitrary files, but now I'm thinking that calculus might have been wrong.

I also think the best solution is to have that simplest solution, that can easily unblock most use cases. Namely the usage of charts in external helmrepos with value files from git.

If arbitrary files were a problem i would think that the repos should be cloned to a single app folder and a chroot used to prevent references outside the scope (app, generator,etc).

**Comment by crenshaw-dev** [comment]
> Whats is the issue with allow to substitute the $ref after the protocol handler secrets//$ref/? That would be the best user expericence.

Path traversal, the bane of the repo-server. :-P

Substituting $ref at the beginning of a valuesFile path is safe and easy, because we know exactly how Helm uses that path, and we can scrub the sensitive, not-to-be-leaked path from error messages.

Substituting in an arbitrary position in the path means that we can no longer be confident that we're successfully scrubbing that path from all error messages or other output.

I think values file substitution should be implemented in a different way: https://github.com/argoproj/argo-cd/pull/11966#issuecomment-1382289061

Instead of substituting the $ref variable with the file's home repo path, we should copy the file out to a new, randomized path and substitute that new, short-lived path into the value file path. We don't care if the path gets leaked, because it'll be deleted by the time anyone sees it.

**Comment by jwitko** [comment]
@crenshaw-dev Based on the very thoughtful and appreciated response in https://github.com/argoproj/argo-cd/pull/12508#issuecomment-1965649318
I'm wondering how that impacts this issue? 
Since this is a bug and not a new feature or feature behavior change is it impacted by the linked decision?
Is there an alternative more localized fix path we could take that wouldn't be impacted by the above decision or is it just time to accept that Helm Secrets will not be a part of multi-source approaches?

**Comment by crenshaw-dev** [comment]
@jwitko I think there's still a pah forward for helm-secrets. I think that something like what I've described at the bottom of this comment https://github.com/argoproj/argo-cd/pull/11966#issuecomment-1382289061 might be viable. I worry about the URL parsing part, but maybe that's not as tricky as it seems in my head.

**Comment by jurgenweber** [comment]
> Soon 2.9.0 will be out. Is there any plans for this? Being able to use encrypted values.yaml is curcial IMHO.
> 
> I am a bit confused, because helm-secrets people mention here that there is a limitation, and this is shown here in this issue. However, there is an example shown which supposes that it should work.
> 
> I am talking about this:
> 
> ##### [Multi-Source Application Support [BETA]](https://github.com/jkroepke/helm-secrets/wiki/ArgoCD-Integration#multi-source-application-support-beta)
> ```
> ArgoCD has limited supported for helm-secrets and Multi-Source application. sops backend is not tested yet.
> 
> References:
> 
> https://github.com/argoproj/argo-cd/issues/11866
> https://github.com/argoproj/argo-cd/pull/11966
> On ArgoCD 2.6.x, helm-secrets isn't supported in Multi-Source application, because the source reference, e.g.: $ref needs to be at the beginn of a string. This is in conflict with helm-secrets, since the string needs to beginn with secrets://. On top, ArgoCD do not resolve references in URLs.
> 
> HELM_SECRETS_VALUES_ALLOW_ABSOLUTE_PATH must be set to true, since ArgoCD pass value files with absolute file path.
> 
> Ensure that the env HELM_SECRETS_WRAPPER_ENABLED=true (default false) and HELM_SECRETS_VALUES_ALLOW_ABSOLUTE_PATH=true is set on the argocd-repo-server. Please ensure you are following the lastest installation instructions (updated on 2023-03-03).
> ```
> 
> and after this text in the same page they provided an example with multi-source application. Could any one tell me what is going on? Or what is the status of this limitation that I don't really understand.

Yeah, this works! Nice find.

**Comment by Maurifc** [comment]
> Soon 2.9.0 will be out. Is there any plans for this? Being able to use encrypted values.yaml is curcial IMHO.
> 
> I am a bit confused, because helm-secrets people mention here that there is a limitation, and this is shown here in this issue. However, there is an example shown which supposes that it should work.
> 
> I am talking about this:
> 
> ##### [Multi-Source Application Support [BETA]](https://github.com/jkroepke/helm-secrets/wiki/ArgoCD-Integration#multi-source-application-support-beta)
> ```
> ArgoCD has limited supported for helm-secrets and Multi-Source application. sops backend is not tested yet.
> 
> References:
> 
> https://github.com/argoproj/argo-cd/issues/11866
> https://github.com/argoproj/argo-cd/pull/11966
> On ArgoCD 2.6.x, helm-secrets isn't supported in Multi-Source application, because the source reference, e.g.: $ref needs to be at the beginn of a string. This is in conflict with helm-secrets, since the string needs to beginn with secrets://. On top, ArgoCD do not resolve references in URLs.
> 
> HELM_SECRETS_VALUES_ALLOW_ABSOLUTE_PATH must be set to true, since ArgoCD pass value files with absolute file path.
> 
> Ensure that the env HELM_SECRETS_WRAPPER_ENABLED=true (default false) and HELM_SECRETS_VALUES_ALLOW_ABSOLUTE_PATH=true is set on the argocd-repo-server. Please ensure you are following the lastest installation instructions (updated on 2023-03-03).
> ```
> 
> and after this text in the same page they provided an example with multi-source application. Could any one tell me what is going on? Or what is the status of this limitation that I don't really understand.

### Thanks. It worked for me.

**This is what I did**

1. Set the variables on `argocd-repo-server` deployment
```yaml
    - name: HELM_SECRETS_VALUES_ALLOW_ABSOLUTE_PATH
      value: "true"
    - name: HELM_SECRETS_WRAPPER_ENABLED
      value: "true"
```

2. Set the encrypted values file path without the **secrets://** prefix (ArgoCD Application)
```
    - helm:
        releaseName: my-app
        valueFiles:
          - $values/secrets.yaml`
 ```

**Comment by jc36** [comment]
This works for me too. Thank you.
But another problem was discovered.
Encrypted with SOPS multiline string (ssh RSA key) **contains a newline** at the end. 
```
gituser_rsa: |
    -----BEGIN OPENSSH PRIVATE KEY-----
    ...
    -----END OPENSSH PRIVATE KEY-----

```
When we use the secrets:// prefix and do not have envvars HELM_SECRETS_VALUES_ALLOW_ABSOLUTE_PATH and HELM_SECRETS_WRAPPER_ENABLED, then newline is passed to the kubernetes secret template (`id_rsa: {{ $.Values.gituser_rsa | b64enc }}`) and secret is mounted to container as a file so everything works as expected.
If we add 
```
    - name: HELM_SECRETS_VALUES_ALLOW_ABSOLUTE_PATH
      value: "true"
    - name: HELM_SECRETS_WRAPPER_ENABLED
      value: "true"
```
then the secrets are also decrypted even without the prefix secrets:// , but it turned out that the value of the variable gets into the secret without the last newline symbol.
So we catch an error reading such an rsa key "Load key "/home/gituser/.ssh/id_rsa": error in libcrypto" and it is clearly related to the missing line translation at the end of the file.
Tried to pass a few additional newline characters at the end and use `|+` (yaml multiline with keeping all newlines at the end), but all newlines are cut off. 

Of course, I have a workaround to add an additional newline in the secret's helm template (`id_rsa: {{ printf "%s\n" $.Values.gituser_rsa | b64enc }}`), but I would like to find the real reason.

I did not find any mention of such a bug in sops "issues".

**Comment by jkroepke** [comment]
Use 

```
gituser_rsa: |-
```

to strip newline at end.

**Comment by jc36** [comment]
We need keep the newline if it exists. 
For some reason, adding the above env variables trims all newlines at the end of decoded variables.

**Comment by stephaneetje** [comment]
Hello,

I'm still unable to use sops backend with GCP KMS on multi source apps. 
Did you all go using GPG to make it work ?

**Comment by stephaneetje** [comment]
I finally managed to make it work, i share here what was wrong in my configuration : https://github.com/jkroepke/helm-secrets/issues/475

**Comment by igoratencompass** [comment]
Hello,

Trying to use Sops for ArgoCD with AWS KMS and `vals` backend. I'm testing a setup where the Application Chart pulls in another chart as dependency in its `Chart.yaml` file.

I use `secrets://` prefix to point to the encrypted file `values-token-enc.yaml` in the `Application`  as per [documentation](https://github.com/jkroepke/helm-secrets/wiki/ArgoCD-Integration) for ArgoCD integration:

```
spec:
  source:
    helm:
      valueFiles:
        - values.yaml
        - secrets://values-token-enc.yaml
    path: path/to/chart
    repoURL: $REPO_URL
    targetRevision: HEAD
```

which is in the same location as the Chart:

```
$ ls path/to/chart
Chart.yaml  values-token-enc.yaml  values.yaml
```

The encrypted file has been generated with the `helm secrets` command and has content:

```
app:
    token:
        clientId: ENC[AES256_GCM,data:...]
        clientToken: ENC[AES256_GCM,data:...]
sops:
    kms:
    -   arn: arn:aws:kms:<region>:<acc_id>:key/<key_id>
...
```

Seems the file gets picked up properly and the values end up in the Chart's `Secret` in the Namespace but the problem is they are not decrypted by ArgoCD! 

The `repo-server` in the ArgoCD Helm chart is properly setup as per [docs](https://github.com/jkroepke/helm-secrets/wiki/ArgoCD-Integration#installation-on-argo-cd) again including the `ServiceAccount` which is annotated with an IAM role that has access to the KMS key granted:

```
  serviceAccount:
    create: true
    name: "argocd-repo-server"
    annotations:
      eks.amazonaws.com/role-arn: arn:aws:iam::<acc_id>:role/<role_name>
    automountServiceAccountToken: true
```

I can't see anything related to Sops in the `repo-server` and `server` logs and the ArgoCD deployment in the UI is fine without any errors. 

Not sure what am I doing wrong or if what I'm trying is supported? Thanks for any help.

**Comment by wofr** [comment]
Maybe it is related to this issue  https://github.com/jkroepke/helm-secrets/wiki/ArgoCD-Integration#multi-source-application-
Its been a while ago I did this setup and its working fine right now, but I do remember I had some similar issues 

Here a code snippet I reference to my encrypted values file
```
 template:
    metadata:
      name: 'bla-{{environment}}'
    spec:
      project: bla-applications
      sources:
      - repoURL: git@github.com:BLA/bla-helm-deployment.git
        targetRevision: "{{releaseHelmVersion}}"
        path: chart
        helm:
          valuesObject:
            global:
              image:
                repo: "eu.gcr.io/bla"
                branch: "{{imageTag}}"
                version: "{{releaseVersion}}"
          valueFiles:
          - $val/bla-environment-configs/{{environmentcfgpath}}/globalsecrets.yaml
          - $val/bla-environment-configs/{{environmentcfgpath}}/globalvalues.yaml
      - repoURL: git@github.com:BLA/bla-argocd-config.git
         targetRevision: "main"
         ref: val          
```

**Comment by igoratencompass** [comment]
Thanks for the quick reply @wofr ! In my case though I have a single source in my application not sure if relevant?

**Comment by wofr** [comment]
Good point! I think you are right.

Maybe you need to explicitly add the kind of encryption you are using like described here (https://github.com/jkroepke/helm-secrets/wiki/ArgoCD-Integration#usage)

E.g ''' secrets+**age**-import:///helm-secrets-private-keys/key.txt?../environmentconfigs/{{environmentcfgpath}}/globalsecrets.yaml'''

...but it seems that should not be necessary in the way you are using it as you use an external key

Have you checked this https://github.com/jkroepke/helm-secrets/wiki/ArgoCD-Integration#external-chart-and-local-values ( https://github.com/argoproj/argo-cd/issues/7257 )

**Comment by igoratencompass** [comment]
@wofr yes I have, I have gone through all the docs. 

The encrypted values file in my case is local and alongside the main `values.yaml` file in the root of the chart as per https://github.com/jkroepke/helm-secrets/wiki/ArgoCD-Integration#external-key-location.

Not sure what's going on but this should just work according to the documentation linked. What I expect to happen is for ArgoCD to decrypt the secrets file via `helm-secrets` plugin and pass it as values file to the `helm template` command that renders the manifests. I don't see the decryption happening though in my case.

EDIT: After cloning the repo on the `repo-server` itself and troubleshooting there running `helm template` it came up I need to use `sops` backend and not `vals`. Problem solved.

**Comment by reggie-k** [comment]
It is recommended to use K8s operators for secret management with GitOps:
https://argo-cd.readthedocs.io/en/stable/operator-manual/secret-management/#secret-management
